core/jvm/src/main/scala/fs2/interop/flow/StreamSubscriber.scala

```diff
@@ -38,230 +38,289 @@ import java.util.concurrent.atomic.AtomicReference
-  /** Called by an upstream reactive-streams system. */
-  override def onError(t: Throwable): Unit = {
+  /** Called by the upstream reactive-streams system when it fails. */
+  override def onError(ex: Throwable): Unit = {
     requireNonNull(
-      t,
+      ex,
       "The throwable provided to onError must not be null"
     )
-    subscriber.onError(t)
+    nextState(Error(ex))
   }
 
-  /** Creates a [[Stream]] from this [[Subscriber]]. */
-  def stream(subscribe: F[Unit]): Stream[F, A] =
-    subscriber.stream(subscribe)
-}
+  /** Called by the upstream reactive-streams system when it has finished sending records. */
+  override def onComplete(): Unit =
+    nextState(Complete(canceled = false))
 
-private[flow] object StreamSubscriber {
+  // Interop API.
 
-  /** Instantiates a new [[StreamSubscriber]] for the given buffer size. */
-  def apply[F[_], A](chunkSize: Int)(implicit F: Async[F]): F[StreamSubscriber[F, A]] = {
-    require(chunkSize > 0, "The buffer size MUST be positive")
-    fsm[F, A](chunkSize).map(fsm => new StreamSubscriber(subscriber = fsm))
+  /** Creates a downstream [[Stream]] from this [[Subscriber]]. */
+  private[flow] def stream(subscribe: F[Unit]): Stream[F, A] = {
+    // Called when downstream has finished consuming records.
+    val finalize =
+      F.delay(nextState(Complete(canceled = true)))
+
+    // Producer for downstream.
+    val dequeue1 =
+      F.async[Option[Chunk[A]]] { cb =>
+        F.delay {
+          nextState(Dequeue(cb))
+
+          Some(finalize)
+        }
+      }
+
+    Stream.bracket(subscribe)(_ => finalize) >>
+      Stream
+        .repeatEval(dequeue1)
+        .unNoneTerminate
+        .unchunks
   }
 
-  /** A finite state machine describing the subscriber. */
-  private[flow] trait FSM[F[_], A] {
+  // Finite state machine.
 
-    /** Receives a subscription from upstream. */
-    def onSubscribe(s: Subscription): Unit
+  /** Helper to reduce noise when creating unary functions. */
+  private def run(block: => Unit): () => Unit = () => block
 
-    /** Receives next record from upstream. */
-    def onNext(a: A): Unit
+  /** Runs a single step of the state machine. */
+  private def step(in: Input[A]): State[A] => (State[A], () => Unit) =
+    in match {
+      case Subscribe(s) => {
+        case Uninitialized(None) =>
+          Idle(s) -> noop
 
-    /** Receives error from upstream. */
-    def onError(t: Throwable): Unit
+        case Uninitialized(Some(cb)) =>
+          WaitingOnUpstream(idx = 0, buffer = null, cb, s) -> run {
+            s.request(chunkSize.toLong)
+          }
+
+        case state =>
+          Failed(new InvalidStateException(operation = "Received subscription", state)) -> run {
+            s.cancel()
+          }
+      }
 
-    /** Called when upstream has finished sending records. */
-    def onComplete(): Unit
+      case Next(a) => {
+        case WaitingOnUpstream(idx, buffer, cb, s) =>
+          implicit val ct = ClassTag[A](a.getClass)
```

Comment:
I don't think this is safe. What happens if `a` is a subclass of `A`, and the subsequent elements are `A` or some different subclass?